pkg/generate/code/compare_helpers.go

```diff
@@ -0,0 +1,460 @@
+
+	specFieldNames := []string{}
+	for fieldName := range r.SpecFields {
+		specField := r.SpecFields[fieldName]
+		if expectHelperFunction(specField.ShapeRef.Shape) {
+		}
+		specFieldNames = append(specFieldNames, fieldName)
+	}
+	sort.Strings(specFieldNames)
+
+	for _, fieldName := range specFieldNames {
+		specField := r.SpecFields[fieldName]
+		memberShapeRef := specField.ShapeRef
+		memberShape := memberShapeRef.Shape
+		shapes := collectSubShapes(memberShape)
+
+		for _, shape := range shapes {
+			shapeGoType := shape.GoTypeWithPkgName()
+			goType := model.ReplacePkgName(shapeGoType, r.SDKAPIPackageName(), "svcapitypes", true)
+			firstFuncParam := "a"
+			secondFuncParam := "b"
+
+			// setting helper function name and signature
+			// func equalRouttingSettingsMap(a, b map[string]*svcapitypes.RouteSettings) bool {
+			out += fmt.Sprintf(
+				"func equal%s(%s, %s %s) bool {\n",
+				shape.ShapeName,
+				firstFuncParam,
+				secondFuncParam,
+				goType,
+			)
+			switch shape.Type {
+			case "map":
+				// Returns Go code that compares two maps
+				out += equalMap(
+					shape,
+					firstFuncParam,
+					secondFuncParam,
+					1,
+				)
+			case "list":
+				// Returns Go code that compares two slices
+				out += equalSlice(
+					shape,
+					firstFuncParam,
+					secondFuncParam,
+					1,
+				)
+			default:
+				msg := "Should not generate a helper function for any type different" +
+					"than string to struct/list/map map and struct/list/map slices"
+				panic(msg)
+			}
+
+			// return true
+			out += fmt.Sprintf("\treturn true\n")
+			// }
+			out += "}\n"
+		}
+	}
+	return out
+}
+
+// expectHelperFunction returns true if a given shape needs a helper function, false
+// otherwise.
+func expectHelperFunction(
+	shape *awssdkmodel.Shape,
+) bool {
+	if shape.Type == "map" {
+		valueType := shape.ValueRef.Shape.Type
+		return valueType == "structure" || valueType == "list" || valueType == "map"
+	}
+
+	if shape.Type == "list" {
+		memberType := shape.MemberRef.Shape.Type
+		return memberType == "structure" || memberType == "list" || memberType == "map"
+	}
+	return false
+}
+
+// collectSubShapes takes a shape and returns all the nested member and value
+// shapes that needs a helper function.
+func collectSubShapes(shape *awssdkmodel.Shape) []*awssdkmodel.Shape {
+	shapes := make([]*awssdkmodel.Shape, 0, 0)
+	if expectHelperFunction(shape) {
+		shapes = append(shapes, shape)
+	}
+
+	switch {
+	case shape.Type == "map":
+		valueShape := shape.ValueRef.Shape
+		switch valueShape.Type {
+		case "map":
+			valueShape := valueShape.ValueRef.Shape
+			shapes = append(shapes, collectSubShapes(valueShape)...)
+		case "list":
+			memberShape := valueShape.MemberRef.Shape
+			shapes = append(shapes, collectSubShapes(memberShape)...)
+		case "structure":
+			for _, memberName := range valueShape.MemberNames() {
+				memberShape := valueShape.MemberRefs[memberName].Shape
+				shapes = append(shapes, collectSubShapes(memberShape)...)
```

Comment:
Should duplicates be taken care of ?

Reply:
Excellent point, i totally missed this. I'm pretty sure we can have duplicates in some cases..